lib/api/objectCopy.js

```diff
@@ -0,0 +1,288 @@
+                    if (headerValResult.error) {
+                        return next(errors.PreconditionFailed);
+                    }
+
+                    if (headerValResult.error) {
+                        log.debug('does not pass conditional copy header test');
+                        return next(headerValResult.error);
+                    }
+                    const storeMetadataParams =
+                        _prepMetadata(sourceObjMD, request.headers,
+                            sourceIsDestination, authInfo, destObjectKey, log);
+                    if (storeMetadataParams.error) {
+                        return next(storeMetadataParams.error);
+                    }
+                    let dataLocator;
+                    // If 0 byte object just set dataLocator to empty array
+                    if (!sourceObjMD.location) {
+                        dataLocator = [];
+                    } else {
+                        // To provide for backwards compatibility before
+                        // md-model-version 2, need to handle cases where
+                        // objMD.location is just a string
+                        dataLocator = Array.isArray(sourceObjMD.location) ?
+                        sourceObjMD.location : [{ key: sourceObjMD.location }];
+                    }
+
+                    if (sourceObjMD['x-amz-server-side-encryption']) {
+                        for (let i = 0; i < dataLocator.length; i++) {
+                            dataLocator[i].masterKeyId = sourceObjMD
+                            ['x-amz-server-side-encryption-aws-kms-key-id'];
+                            dataLocator[i].algorithm =
+                                sourceObjMD['x-amz-server-side-encryption'];
+                        }
+                    }
+                    return next(null, storeMetadataParams, dataLocator);
+                });
+        },
+        function checkDestAuth(storeMetadataParams, dataLocator, next) {
+            return services.metadataValidateAuthorization(valPutParams,
+                (err, destBucketMD, destObjMD) => {
+                    if (err) {
+                        log.debug('error validating put part of request',
+                        { error: err });
+                        return next(err);
+                    }
+                    const flag = destBucketMD.hasDeletedFlag()
+                        || destBucketMD.hasTransientFlag();
+                    if (flag) {
+                        log.trace('deleted flag or transient flag ' +
+                        'on destination bucket', { flag });
+                        return next(errors.NoSuchBucket);
+                    }
+                    return next(null, storeMetadataParams,
+                        dataLocator, destBucketMD, destObjMD);
+                });
+        },
+        function getCipherBundle(storeMetadataParams, dataLocator, destBucketMD,
```

Comment:
getting a cipher bundle implies a network communication with the KMS. It may be best amortized close to the data.put. e.g. data transfers starting independently of other parts being ready regarding there cipherbundle